src/file-events/cpp/generic_fsnotifier.cpp

```diff
@@ -163,9 +164,55 @@ void AbstractServer::reportChange(JNIEnv* env, int type, const u16string& path)
     jstring javaPath = env->NewString((jchar*) path.c_str(), (jsize) path.length());
     env->CallVoidMethod(watcherCallback, watcherCallbackMethod, type, javaPath);
     env->DeleteLocalRef(javaPath);
+
+    jthrowable exception = env->ExceptionOccurred();
+    if (exception != nullptr) {
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+
+        jclass exceptionClass = env->GetObjectClass(exception);
+        assert(exceptionClass != nullptr);
+        jmethodID getMessage = env->GetMethodID(exceptionClass, "getMessage", "()Ljava/lang/String;");
+        assert(getMessage != nullptr);
+        jstring javaMessage = (jstring) env->CallObjectMethod(exception, getMessage);
+        assert(javaMessage != nullptr);
+        string message = javaToUtf8String(env, javaMessage);
+        env->DeleteLocalRef(javaMessage);
+
+        jclass classClass = env->FindClass("java/lang/Class");
+        assert(classClass != nullptr);
+        jmethodID getClassName = env->GetMethodID(classClass, "getName", "()Ljava/lang/String;");
+        assert(getClassName != nullptr);
+        jstring javaExceptionType = (jstring) env->CallObjectMethod(exceptionClass, getClassName);
+        assert(javaExceptionType != nullptr);
+        string exceptionType = javaToUtf8String(env, javaExceptionType);
+        env->DeleteLocalRef(javaExceptionType);
+
+        env->DeleteLocalRef(classClass);
+        env->DeleteLocalRef(exceptionClass);
+        env->DeleteLocalRef(exception);
+
+        throw FileWatcherException("Caught " + exceptionType + " while calling callback: " + message);
+    }
+}
+
+void AbstractServer::reportError(JNIEnv* env, const exception& exception) {
+    jclass exceptionClass = env->FindClass("net/rubygrapefruit/platform/NativeException");
+    assert(exceptionClass != nullptr);
+    u16string message = utf8ToUtf16String(exception.what());
+    jstring javaMessage = env->NewString((jchar*) message.c_str(), (jsize) message.length());
+    jmethodID constructor = env->GetMethodID(exceptionClass, "<init>", "(Ljava/lang/String;)V");
+    jobject javaException = env->NewObject(exceptionClass, constructor, javaMessage);
+    assert(javaException != nullptr);
+    env->CallVoidMethod(watcherCallback, watcherReportErrorMethod, javaException);
+    env->DeleteLocalRef(exceptionClass);
```

Comment:
Do you also need to delete the other local refs to `javaMessage`, `constructor` and `javaException`?

Reply:
Ah, I missed a couple things here, fix: https://github.com/gradle/native-platform/pull/126.